### ISSUE TITLE
Triggers that share common prefix cannot co-exist

### ISSUE THREAD

**Issue by jrop** [federico-terzi/espanso#178]
I have two triggers that I want to set up: I want each to expand to a different email:

```
  - trigger: ":email"
    replace: "a@b.c"
  - trigger: ":email2"
    replace: "d@e.f"
```

However, when I type `:email`, espanso does not wait to see if there is a `2` coming: it eagerly expands `:email` to `a@b.c`.

I propose the following solution:
* Introduce a new "timeout" config value (a decent default may be 500ms)
* Change analysis of triggers in a file to build up a "match" state machine.
* At any terminal node in the state machine that has the potential to eat more characters, start the stopwatch (configurable): if the timer finished before there is more input, then insert the match corresponding to the current node of the state machine.  Otherwise, continue state transitions.

**Comment by federico-terzi** [comment]
Hey @jrop,
Thank you for the suggestion, this feature could be useful indeed.
I'll try to make a few experiments and get back to you when I got something!

Thanks for your help :)

**Comment by AryToNeX** [comment]
imho since the time-based approach inevitably makes you wait to write the rest of the phrase (and it can be frustrating to some) I would propose
- Key combination approach: pressing a certain key (combination) will make espanso "expand" the longest match it could find right before the key (combination) was pressed. I think of it as a short-term passive mode where espanso would store the phrase starting from the last typed pattern, then it would process it in RAM, it would go all the way back and then it would paste the new processed sentence.
- Spacebar approach: if two matches are the same and they don't contain spaces in them, then pressing space will make espanso do its job on the longest match it could find right before the spacebar was pressed

**Comment by jrop** [comment]
Waiting for future states only needs to happen, though, if there is a possibility of future input. One can prune out that branch of the state machine if the node you are currently on is a terminal node. This is how vim does its key bindings. Anytime a key binding is used that has no possibility of more future keys, it takes effect immediately.

**Comment by federico-terzi** [comment]
Hey guys,
Both your suggestions are equally valid. I would personally go with the first one proposed by @jrop as it is the most seamless one. That said, I'm a bit worried about the potential delays of that approach. For that reason, I would probabily disable it by default, and allow the users to enable it manually.

Also because personally, rather than having `:email` and `:email2`, I would have something like:
* `:email1` and `:email2`
* `:email` and `:2email`

But I can understand why users would prefer the timeout approach, so I'm definitely planning to add the feature.

Cheers

**Comment by stheid** [comment]
I guess with the "word" flag enabled this is not that severe, right, as than, the trigger is only executed after a next whitespace has been hit.

**Comment by federico-terzi** [comment]
> I guess with the "word" flag enabled this is not that severe, right, as than, the trigger is only executed after a next whitespace has been hit.

@stheid I agree with you. Using the `word` property solves many of these problems, therefore making this issue less critical.

**Comment by SpecialCharacter** [comment]
The issue is also the "philosophy" of espanso: Do you put the "activation" at the beginning or at the end?
Having to enter a lot of special characters in AutoHotkey, I settled for the latter.
So a:# --> ä, a:-# --> ǟ without messing up the cases where I actually want a: or ä-.
In your case, it would be email: or email2:, respectively.
Is this possible with a (future version of) espanso?

**Comment by federico-terzi** [comment]
@SpecialCharacter Thank you for the feedback!

> Is this possible with a (future version of) espanso?

Isn't this already possible? :) 

I mean, if you define triggers as `a:#` and `a:-#` espanso should be able to handle them out of the box, as neither of them is a prefix of the other.
As you said, if you place a different "activation char" at the end, espanso is able to differentiate between the two, as long as neither of them is a prefix of the other

**Comment by SpecialCharacter** [comment]
I was referring to jrop's question.
If it works already now, even better! :)
